### PR TITLE
Edit users

### PR DESCRIPTION
**Pull request by kamilsobik**
# Edit users
- [x] changing table with users
- [x] add modal with edit one user - changing password, email, account type (only admin can do this)
- [x] changing user model - creating userType (admin, requestor, responder)
- [x] save data to user profile
- [x] change permissions - admin can view all routes, requestor and responder view only allowed routes
- [x] add helper to add row color depending of user type
- [x] changing button to shared button
- [x] add function to delete user
- [x] add Ember Power Select and implement as shared component
- [x] add ember changeset in user edit modal

https://user-images.githubusercontent.com/99078349/202846766-b9260c19-9e37-4c48-a5ec-c3d61111cefd.mov

### CODE REVIEW

File: app/application/route.js [hunk 1]
```diff
@@ -16,26 +16,23 @@ export default class ApplicationRoute extends Route {
       username: 'admin',
       password: 'admin123',
       email: 'admin@gmail.com',
-      isRequestor: true,
-      isAdmin: true,
+      userType: 'admin',
```
Comment (wozny1989) [inline app/application/route.js:19]: ❤️

File: app/components/modal/edit-user/component.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+  @action
+  onPasswordChange(event) {
+    this.args.user.password = event.target.value;
+  }
+
+  @action
+  onEmailChange(event) {
+    this.args.user.email = event.target.value;
+  }
```
Comment (wozny1989) [inline app/components/modal/edit-user/component.js]: How about to do that as you did above?
```suggestion
  @action
  onPasswordChange({ target: { value }}) {
    this.args.user.password = value;
  }

  @action
  onEmailChange({ target: { value }}) {
    this.args.user.email = value;
  }
```

File: app/components/modal/edit-user/template.hbs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        <div class='mb-3'>
+          <label for='emailUserInput' class='form-label'>Enter new email</label>
+          <input
+            type='e-mail'
+            class='form-control'
+            id='editEmailInput'
+            placeholder='{{@user.email}}'
+            value={{this.email}}
+            {{on 'input' this.onEmailChange}}
+          />
```
Comment (wozny1989) [inline app/components/modal/edit-user/template.hbs]: How about to create shared input component in this PR? Which by default will have `type='text'` and `class='form-control'`. Example invoke in this case:
```hbs
<Shared::Input
  @id='editEmailInput'
  @type='e-mail'
  @placeholder={{@user.email}}
  @value={{this.email}}
  @onChange={{this.onEmailChange}}
/>
```

File: app/components/modal/edit-user/template.hbs [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        </div>
+        <div class='mb-3'>
+          <label for='passwordUserInput' class='form-label'>Enter new password</label>
+          <input
+            type='password'
+            class='form-control'
+            id='editPasswordInput'
+            placeholder='{{@user.password}}'
```
Comment (wozny1989) [inline app/components/modal/edit-user/template.hbs]: Are not necessary
```suggestion
            placeholder={{@user.password}}
```

File: app/components/modal/edit-user/template.hbs [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+            value={{this.password}}
+            {{on 'input' this.onPasswordChange}}
+          />
+        </div>
+        <div class='mb-3'>
+          <label for='userTypeInput' class='form-label'>Change user type:</label>
+          <div {{on 'input' (fn this.onUserTypeChange 'userType')}}>
```
Comment (wozny1989) [inline app/components/modal/edit-user/template.hbs]: Linter not shout about add `role="button"`? 🤔 https://github.com/ember-template-lint/ember-template-lint/issues/41

File: app/components/modal/edit-user/template.hbs [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+        />
+        <Shared::Button
+          class='btn-danger'
+          @label='Cancel'
+          @onClick={{this.onCancel}}
+        />
+        <Shared::Button
+          class='btn-success'
+          @label='Save change'
+          @onClick={{this.onSaveChange}}
```
Comment (wozny1989) [inline app/components/modal/edit-user/template.hbs]: 😍

File: app/components/navbar/component.js [hunk 7]
```diff
@@ -1,10 +1,34 @@
...
+        this.session.currentUser.userType === 'admin'
+    );
+  }
+
+  get isUserTypeIsResponder() {
+    return Boolean(
+      this.session.currentUser.userType === 'responder' ||
+        this.session.currentUser.userType === 'admin'
+    );
+  }
```
Comment (wozny1989) [inline app/components/navbar/component.js]: 🤔 
```suggestion
  get isUserTypeIsRequestor() {
    return Boolean(
      this.session.currentUser.userType === 'requestor' ||
        this.isUserTypeIsAdmin
    );
  }

  get isUserTypeIsResponder() {
    return Boolean(
      this.session.currentUser.userType === 'responder' ||
        this.isUserTypeIsAdmin
    );
  }
```

File: app/components/navbar/template.hbs [hunk 8]
```diff
@@ -19,15 +19,21 @@
...
+        {{#if this.isUserTypeIsResponder}}
+          <li class='nav-item'>
+            <LinkTo class='nav-link' @route='home.close-request'>Close request</LinkTo>
+          </li>
+        {{/if}}
+        {{#if this.isUserTypeIsAdmin}}
+          <li class='nav-item'>
+            <LinkTo class='nav-link' @route='home.users'>Users</LinkTo>
+          </li>
+        {{/if}}
```
Comment (wozny1989) [inline app/components/navbar/template.hbs]: How about change naming for better readability?
```suggestion
        {{#if this.isRequestorUserType}}
          <li class='nav-item'>
            <LinkTo class='nav-link' @route='home.open-request'>Open new request</LinkTo>
          </li>
        {{/if}}
        {{#if this.isResponderUserType}}
          <li class='nav-item'>
            <LinkTo class='nav-link' @route='home.close-request'>Close request</LinkTo>
          </li>
        {{/if}}
        {{#if this.isAdminUserType}}
          <li class='nav-item'>
            <LinkTo class='nav-link' @route='home.users'>Users</LinkTo>
          </li>
        {{/if}}
```

File: app/components/user-table/user-table-row/component.js [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+import { tracked } from '@glimmer/tracking';
+
+export default class UserTableUserTableRow extends Component {
+  @tracked isShowUserModal = false;
+
+  @action
+  onShowUserModal() {
+    this.isShowUserModal = true;
+  }
+  @action
```
Comment (wozny1989) [inline app/components/user-table/user-table-row/component.js]: Please, add empty line between

File: app/helpers/class-user-table.js [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+    newTableClass = 'table-success';
+  } else if (userType === 'reqponder') {
+    newTableClass = 'table-warning';
+  } else if (userType === 'admin') {
+    newTableClass = 'table-danger';
+  }
+  return newTableClass;
+}
+
+export default helper(classUserTable);
```
Comment (wozny1989) [inline app/helpers/class-user-table.js]: Small tip for this condition case, what do you think about that?:
```suggestion
import { helper } from '@ember/component/helper';

const tableClass = {
  requestor: 'table-success',
  responder: 'table-warning',
  admin: 'table-danger'
}

export function classUserTable([userType]) {
  return tableClass[userType] || 'table-primary';
}

export default helper(classUserTable);
```

File: app/helpers/class-user-table.js [hunk 11]
```diff
@@ -0,0 +1,15 @@
+import { helper } from '@ember/component/helper';
+
+export function classUserTable([userType]) {
+  let newTableClass = 'table-primary';
+  if (userType === 'requestor') {
+    newTableClass = 'table-success';
+  } else if (userType === 'reqponder') {
```
Comment (wozny1989) [inline app/helpers/class-user-table.js]: typo
```suggestion
  } else if (userType === 'responder') {
```

File: app/register/template.hbs [hunk 12]
```diff
@@ -41,10 +41,11 @@
             <label class='form-label' for='password'>Password</label>
           </div>
           <div class='text-center text-lg-start mt-4 pt-2'>
-            <button
-              type='submit'
-              class='btn btn-primary btn-lg button-login'
-            >Register new user</button>
+            <Shared::Button
+              class='btn-primary btn-lg button-login'
```
Comment (wozny1989) [inline app/register/template.hbs]: Maybe it's time to add additional property to your Shared::Button like `@size="lg"` (or `"sm"`) which will inject this extra class?

**Review by wozny1989 — CHANGES_REQUESTED** [review]
Few things but overall nice work, I see you feel it!!! 🕺

File: app/components/modal/edit-user/template.hbs [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+          <div {{on 'input' (fn this.onUserTypeChange 'userType')}}>
+            <select
+              class='form-select'
+              aria-label='Default select example'
+              value={{@user.userType}}
+            >
+              <option value='requestor'>Requestor</option>
+              <option value='responder'>Responder</option>
+              <option value='admin'>Admin</option>
+            </select>
```
Comment (DawidDamek) [inline app/components/modal/edit-user/template.hbs]: Maybe try to rewrite it to powerselect🤔 Wodzu fixed the documentation so it should be doable

File: app/components/navbar/component.js [hunk 14]
```diff
@@ -1,10 +1,34 @@
...
 import { inject as service } from '@ember/service';
 
 export default class NavbarComponent extends Component {
+  @storageFor('logged-as') loggedAs;
   @service session;
 
+  get loginUserUsername() {
+    return this.loggedAs.get('username');
+  }
+
```
Comment (DawidDamek) [inline app/components/navbar/component.js:13]: Dont you have username in `this.session.currentUser.username`? If you have it there you can remove one import

File: app/components/shared/button/template.hbs [hunk 15]
```diff
@@ -3,4 +3,7 @@
   type={{this.type}}
   {{on 'click' this.onClick}}
   ...attributes
->{{@label}}</button>
\ No newline at end of file
+>
+  {{#if @faIcon}} <FaIcon @icon={{@faIcon}} />{{/if}}
+  {{#if @src}} <img src={{@src}} alt={{@alt}} />{{/if}}
+  {{@label}}</button>
```
Comment (DawidDamek) [inline app/components/shared/button/template.hbs:9]: 👌

File: app/session/service.js [hunk 16]
```diff
@@ -20,6 +20,8 @@ export default class SessionService extends Service {
     if (isUserExist) {
       const user = users.firstObject;
       this.loggedAs.set('id', user.id);
+      this.loggedAs.set('username', user.username);
+      this.loggedAs.set('userType', user.userType);
```
Comment (DawidDamek) [inline app/session/service.js]: 🤔 why do you set it there?

File: app/components/providers/open-request/template.hbs [hunk 17]
```diff
@@ -49,8 +49,9 @@
   </div>
   <div class='d-grid gap-2'>
     <Shared::Button
-      class='btn-success btn-lg mt-5'
+      class='btn-success mt-5'
       @label='Create request'
+      @size='btn-lg'
```
Comment (wozny1989) [inline app/components/providers/open-request/template.hbs]: How about:
```suggestion
      @size='lg'
```

and make getter inside shared component something like and concat to basic class:
```js
get size() {
  const { size } = this.args
  const isCorrectSize = ['sm', 'lg'].includes(size);
  return isCorrectSize ? `btn-${size}` : '';
}
```

**Review by wozny1989 — CHANGES_REQUESTED** [review]

File: app/utils/form-data.js [hunk 18]
```diff
@@ -0,0 +1,3 @@
+const userType = ['admin', 'responder', 'requestor'];
+
+export { userType };
```
Comment (wozny1989) [inline app/utils/form-data.js]: It should be more readable, and easier to know that it's a collection:
```suggestion
const userTypes = ['admin', 'responder', 'requestor'];

export { userTypes };
```

File: app/components/modal/edit-user/component.js [hunk 19]
```diff
@@ -0,0 +1,39 @@
+import Component from '@glimmer/component';
+import { action } from '@ember/object';
+import { userType } from 'mms-ember-project/utils/form-data';
+
+export default class ModalEditUserComponent extends Component {
+  userType = userType;
```
Comment (wozny1989) [inline app/components/modal/edit-user/component.js]: Nice idea with this import! ❤️ One thing:
```suggestion
import { userTypes } from 'mms-ember-project/utils/form-data';

export default class ModalEditUserComponent extends Component {
  userTypes = userTypes;
```

File: app/components/modal/edit-user/template.hbs [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+      <div class='modal-body'>
+        <div class='mb-3'>
+          <label for='emailUserInput' class='form-label'>Enter new email</label>
+          <Shared::Input
+            @type='email'
+            @id='editEmailInput'
+            @placeholder={{@user.email}}
+            @value={{this.email}}
+            @onChange={{this.onEmailChange}}
+          />
```
Comment (wozny1989) [inline app/components/modal/edit-user/template.hbs]: Shared component! 💪 
But whats mean `this.email`? I didn't see that property in `component.js`. If you want to display original user email record in `placeholder`, consider to make something like a 'snapshot' this property in constructor and create getter for `this.email`:
```js
constructor() {
  super(...arguments);
  this.originalEmail = this.args.user.email;
}

@computed('args.user.email')
get email() {
  const { lang } = this.args.user.changedAttributes();
  return lang?.[1] || '';
}
```

then:
```suggestion
          <Shared::Input
            @type='email'
            @id='editEmailInput'
            @placeholder={{this.originalEmail}}
            @value={{this.email}}
            @onChange={{this.onEmailChange}}
          />
```

File: app/components/modal/edit-user/template.hbs [hunk 21]
```diff
@@ -0,0 +1,59 @@
...
+        </div>
+        <div class='mb-3'>
+          <label for='passwordUserInput' class='form-label'>Enter new password</label>
+          <Shared::Input
+            @type='password'
+            @id='editPasswordInput'
+            @placeholder={{@user.password}}
+            @value={{this.password}}
+            @onChange={{this.onPasswordChange}}
+          />
```
Comment (wozny1989) [inline app/components/modal/edit-user/template.hbs]: same here ☝️

File: app/register/template.hbs [hunk 22]
```diff
@@ -41,10 +41,12 @@
...
-            <button
-              type='submit'
-              class='btn btn-primary btn-lg button-login'
-            >Register new user</button>
+            <Shared::Button
+              class='btn-primary button-login'
+              @size='lg'
+              @label='Register new user'
+              @type='subtmit'
+            />
```
Comment (wozny1989) [inline app/register/template.hbs]: How about create another property, hmm like `variant`?
```suggestion
            <Shared::Button
              class='button-login'
              @variant='primary'
              @size='lg'
              @label='Register new user'
              @type='subtmit'
            />
```

File: app/helpers/class-user-table.js [hunk 23]
```diff
@@ -0,0 +1,10 @@
+import { helper } from '@ember/component/helper';
+const tableClass = {
+  requestor: 'table-success',
+  responder: 'table-warning',
+  admin: 'table-danger',
+};
+export function classUserTable([userType]) {
+  return tableClass[userType] || 'table-primary';
+}
+export default helper(classUserTable);
```
Comment (wozny1989) [inline app/helpers/class-user-table.js]: 💪 
+ empty lines 🙏 
```suggestion
import { helper } from '@ember/component/helper';
const tableClass = {
  requestor: 'table-success',
  responder: 'table-warning',
  admin: 'table-danger',
};

export function classUserTable([userType]) {
  return tableClass[userType] || 'table-primary';
}

export default helper(classUserTable);
```

File: app/helpers/class-table.js [hunk 24]
```diff
@@ -1,15 +1,11 @@
 import { helper } from '@ember/component/helper';
+const tableClass = {
+  '1.0 tpm': 'table-success',
+  '2.0 fault': 'table-warning',
+  '3.0 breakdown': 'table-danger',
+};
```
Comment (wozny1989) [inline app/helpers/class-table.js]: Keys with dots and white space look weird... Consider to change structure:
```suggestion
const tableClasses = [
  { value: '1.0 tpm', class: 'table-success' },
  { value: '2.0 fault', class: 'table-warning' },
  { value: '3.0 breakdown', class: 'table-danger' }
];
```

File: app/helpers/class-table.js [hunk 25]
```diff
@@ -1,15 +1,11 @@
...
-  let newTableClass = 'table-primary';
-  if (typeRequest === '1.0 tpm') {
-    newTableClass = 'table-success';
-  } else if (typeRequest === '2.0 fault') {
-    newTableClass = 'table-warning';
-  } else if (typeRequest === '3.0 breakdown') {
-    newTableClass = 'table-danger';
-  }
-  return newTableClass;
+  return tableClass[typeRequest] || 'table-primary';
```
Comment (wozny1989) [inline app/helpers/class-table.js]: then:
```suggestion
  return tableClass.find(({ value }) => value ===  typeRequest).class || 'table-primary';
```

File: app/components/shared/input/template.hbs [hunk 26]
```diff
@@ -0,0 +1,8 @@
+<input
+  type={{this.chceckInputType}}
+  class='form-control {{@class}} {{this.size}}'
+  id={{@id}}
+  placeholder='{{@placeholder}}'
+  value={{@vale}}
+  {{on 'input' @onChange}}
+/>
```
Comment (wozny1989) [inline app/components/shared/input/template.hbs]: Add `...attributes` then you can remove Ember `@class` attribute, then you can pass natively `class` prop, then will be concat other classes to this element (similar like you have in shared/button):
```suggestion
<input
  type={{this.chceckInputType}}
  class='form-control {{this.size}}'
  id={{@id}}
  placeholder='{{@placeholder}}'
  value={{@vale}}
  {{on 'input' @onChange}}
  ...attributes
/>
```

**Review by wozny1989 — CHANGES_REQUESTED** [review]

**Comment by wozny1989** [comment]
@kamilsobik Additional thing, bad label - `password` instead `email`:
image

File: app/components/modal/edit-user/component.js [hunk 27]
```diff
@@ -0,0 +1,43 @@
+import Component from '@glimmer/component';
+import { action } from '@ember/object';
+import { Changeset } from 'ember-changeset';
+import { userTypes } from 'mms-ember-project/utils/form-data';
+
+export default class ModalEditUserComponent extends Component {
+  userTypes = userTypes;
+
+  changeset = Changeset(this.args.user);
```
Comment (wozny1989) [inline app/components/modal/edit-user/component.js]: it's working but it's not perfect way - consider ot move this to `contructor` like in this example https://github.com/poteto/ember-changeset#types then will be perfect ❤️

File: app/components/modal/show-request/template.hbs [hunk 28]
```diff
@@ -8,12 +8,12 @@
...
-            class='form-control'
-            id='downtimeInput'
-            value={{this.downtime}}
-            {{on 'input' this.onDowntimeChange}}
+          <Shared::Input
+            @type='number'
+            @id='downtimeInput'
+            @placeholder={{@user.password}}
+            @value={{this.downtime}}
+            @onChange={{this.onDowntimeChange}}
```
Comment (wozny1989) [inline app/components/modal/show-request/template.hbs]: 💪

File: app/components/shared/power-select/template.hbs [hunk 29]
```diff
@@ -0,0 +1,10 @@
+<PowerSelect
+  @triggerClass='form-control'
+  @selected={{@selected}}
+  @options={{@options}}
+  @onChange={{@onChange}}
+  @placeholder={{@placeholder}}
+  as |value|
+>
+  {{value}}
+</PowerSelect>
```
Comment (wozny1989) [inline app/components/shared/power-select/template.hbs]: Almost done! 😄 But what's happen if you'll have other object structure and for proper display you have to use nested key, like:
```hbs
<PowerSelect
  @triggerClass='form-control'
  @selected={{@selected}}
  @options={{@options}}
  @onChange={{@onChange}}
  @placeholder={{@placeholder}}
  as |value|
>
  {{value.name}}
</PowerSelect>
```

I suggest to change:
```suggestion
<PowerSelect
  @triggerClass='form-control'
  @selected={{@selected}}
  @options={{@options}}
  @onChange={{@onChange}}
  @placeholder={{@placeholder}}
  as |value|
>
  {{yield value}}
</PowerSelect>
```

Then using of this component should be always via block type:
```hbs
<Shared::PowerSelect
  @selected={{this.record.machine}}
  @options={{this.machines}}
  @onChange={{fn this.onSelectionChange 'machine'}}
  @placeholder={{'Select machine'}}
  as |value|
>
  {{value}}
</Shared::PowerSelect>
```

or if we have nested structure:
```hbs
<Shared::PowerSelect
  @selected={{this.record.machine}}
  @options={{this.machines}}
  @onChange={{fn this.onSelectionChange 'machine'}}
  @placeholder={{'Select machine'}}
  as |value|
>
  {{value.name}}
</Shared::PowerSelect>
```

File: app/components/user-table/user-table-row/template.hbs [hunk 30]
```diff
@@ -0,0 +1,18 @@
...
+      class='btn'
+      @variant='success'
+      @faIcon='user-pen'
+      @onClick={{this.onShowUserModal}}
+    />
+  </td>
+</tr>
+
+{{#if this.isShowUserModal}}
+  <Modal::EditUser @onClose={{this.onHideUserModal}} @user={{@user}} />
```
Comment (wozny1989) [inline app/components/user-table/user-table-row/template.hbs]: 🔄 
```suggestion
  <Modal::EditUser @user={{@user}} @onClose={{this.onHideUserModal}}/>
```

File: app/components/providers/open-request/template.hbs [hunk 31]
```diff
@@ -1,57 +1,39 @@
...
-          <option value='3.0 breakdown'>3.0 Breakdown</option>
-        </select>
-      </div>
+    <div class='col mb-5'>
+      <Shared::PowerSelect
+        @selected={{this.record.type}}
+        @options={{this.requestTypes}}
+        @onChange={{fn this.onTypeSelectionChange 'type'}}
+        @placeholder={{'Select request type'}}
+      />
```
Comment (wozny1989) [inline app/components/providers/open-request/template.hbs]: https://gist.github.com/wozny1989/30c4c3e2e6a86d773366de933997edbe#file-example_template_refactoring-hbs-L166-L183
```suggestion
      <Shared::PowerSelect
        @placeholder={{'Select request type'}}
        @selected={{this.record.type}}
        @options={{this.requestTypes}}
        @onChange={{fn this.onTypeSelectionChange 'type'}}
      />
```

File: app/components/providers/open-request/template.hbs [hunk 32]
```diff
@@ -1,57 +1,39 @@
...
-          <option value='Machine 3'>Machine 3</option>
-        </select>
-      </div>
+    <div class='col mb-5'>
+      <Shared::PowerSelect
+        @selected={{this.record.machine}}
+        @options={{this.machines}}
+        @onChange={{fn this.onSelectionChange 'machine'}}
+        @placeholder={{'Select machine'}}
+      />
```
Comment (wozny1989) [inline app/components/providers/open-request/template.hbs]: same here ⬆️

File: app/components/shared/button/component.js [hunk 33]
```diff
@@ -1,11 +1,33 @@
...
+      'danger',
+      'warning',
+      'info',
+      'light',
+      'dark',
+      'link',
+    ].includes(variant);
+    return isCorrectVariant ? `btn-${variant}` : '';
+  }
+
```
Comment (wozny1989) [inline app/components/shared/button/component.js:30]: 🚀

File: app/components/shared/button/template.hbs [hunk 34]
```diff
@@ -1,6 +1,9 @@
 <button
-  class='btn'
+  class='btn {{this.size}} {{this.variant}}'
```
Comment (wozny1989) [inline app/components/shared/button/template.hbs]: How about create another getter like `buttonClasses`?
```js
get buttonClasses() {
  const classes = ['btn'];
  const { size, variant } = this;
  
  if (size) {
    classes.push(size);
  }

  if (variant) {
    classes.push(variant);
  }

  return classes.join(' ');
}
```

```suggestion
  class={{this.buttonClasses}}
```

File: app/components/shared/input/template.hbs [hunk 35]
```diff
@@ -0,0 +1,9 @@
+<input
+  type={{this.chceckInputType}}
+  class='form-control {{this.size}}'
```
Comment (wozny1989) [inline app/components/shared/input/template.hbs]: consider to create similar class as I suggested in `shared/button`

File: app/components/shared/input/template.hbs [hunk 36]
```diff
@@ -0,0 +1,9 @@
...
+  id={{@id}}
+  placeholder='{{@placeholder}}'
+  value={{@vale}}
```
Comment (wozny1989) [inline app/components/shared/input/template.hbs]:
```suggestion
  value={{@value}}
```

File: app/components/shared/input/template.hbs [hunk 37]
```diff
@@ -0,0 +1,9 @@
...
+  {{on 'input' @onChange}}
```
Comment (wozny1989) [inline app/components/shared/input/template.hbs]: How about create action inside this component, and pass more comfortable `value` attribute, instead every time taking `event.target.value`. 
Example:
```suggestion
  {{on 'input' this.onChange}}
```

```js
// shared/input/component.js
@action
onChange(event) {
  const { value } = event.target;
  this.args.onChange?.(value, event);
}
```

Then when you'll using this shared component you'll have lovely parameters:
```hbs
<Shared::Input 
  @value={{this.value}}
  @onChange={{this.onChange}}
/>
```

```js
@action
onChange(value, /* event */) {
  this.value = value; // If you want to use event, you have it in second argument 🚀
}
```

**Review by wozny1989 — CHANGES_REQUESTED** [review]

File: app/components/shared/power-select/template.hbs [hunk 38]
```diff
@@ -6,5 +6,5 @@
   @placeholder={{@placeholder}}
   as |value|
 >
-  {{value}}
+  {{yield value}}
```
Comment (wozny1989) [inline app/components/shared/power-select/template.hbs:9]: ❤️

**Review by wozny1989 — APPROVED** [review]
Nice, I really like it!

**Pull request merged**